include/crocoddyl/multibody/costs/centroidal-momentum.hxx

```diff
@@ -93,15 +86,19 @@ template <typename Scalar>
 void CostModelCentroidalMomentumTpl<Scalar>::set_referenceImpl(const std::type_info& ti, const void* pv) {
   if (ti == typeid(Vector6s)) {
     href_ = *static_cast<const Vector6s*>(pv);
+    ResidualModelCentroidalMomentum* residual = static_cast<ResidualModelCentroidalMomentum*>(residual_.get());
+    residual->set_reference(href_);
   } else {
     throw_pretty("Invalid argument: incorrect type (it should be Vector6s)");
   }
 }
 
 template <typename Scalar>
-void CostModelCentroidalMomentumTpl<Scalar>::get_referenceImpl(const std::type_info& ti, void* pv) const {
+void CostModelCentroidalMomentumTpl<Scalar>::get_referenceImpl(const std::type_info& ti, void* pv) {
   if (ti == typeid(Vector6s)) {
     Eigen::Map<Vector6s> ref_map(static_cast<Vector6s*>(pv)->data());
+    ResidualModelCentroidalMomentum* residual = static_cast<ResidualModelCentroidalMomentum*>(residual_.get());
+    href_ = residual->get_reference();
     ref_map[0] = href_[0];
     ref_map[1] = href_[1];
     ref_map[2] = href_[2];
```

Comment:
It looks like you are 
-> first copying residual->href_ into href_,
-> and then copying href_ into ref_map.

Instead, I suggest to make href_ as a reference that binds to residual ->href_ at construction time.
And to assign this to the pv pointer, use (syntax may not be correct)
```c++
pv = &residual->href_(0); // get the address storing the data for href_
Eigen::Map<Vector6s> ref_map (pv, 6); 
```

Reply:
This code will be deprecated in the follow-up PR (very soon).

I will avoid any unnecessary effort on improving the tailored cost functions.